### PR TITLE
mqtt: support for tasmota power ports

### PR DESCRIPTION
**Pull request by Emantor**
**Description**
Support for MQTT Tasmota Devices.

**Checklist**
- [x] Documentation for the feature
- [x] Tests for the feature 
- [x] The arguments and description in doc/configuration.rst have been updated
- [ ] CHANGES.rst has been updated
- [x] PR has been tested

### CODE REVIEW

File: labgrid/driver/mqtt.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+@target_factory.reg_driver
+@attr.s(eq=False)
+class TasmotaPowerDriver(Driver, PowerProtocol):
+    bindings = {
+            "power": {"TasmotaPowerPort", "NetworkTasmotaPowerPort"}
+    }
+    delay = attr.ib(default=2.0, validator=attr.validators.instance_of(float))
+    _client = attr.ib(default=mqtt.Client(), validator=attr.validators.instance_of(mqtt.Client))
+    _status_lock = attr.ib(default=threading.Lock())
```
Comment (jluebbe) [inline labgrid/driver/mqtt.py:27]: Either initialize in `__attrs_post_inst__` or use `factory=`.

File: labgrid/resource/lxaiobus.py [hunk 2]
```diff
@@ -10,13 +10,14 @@
...
+
+    _last = attr.ib(default=0.0, validator=attr.validators.instance_of(float))
+
     def __attrs_post_init__(self):
         super().__attrs_post_init__()
 
         self.log = logging.getLogger('LXAIOBusNodeManager')
 
-        self._last = 0.0
-
```
Comment (jluebbe) [inline labgrid/resource/lxaiobus.py:19]: These changes are unrelated.

File: labgrid/resource/mqtt.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+from ..factory import target_factory
+from ..util import Timeout
+
+@attr.s(eq=False)
+class MQTTManager(ResourceManager):
+    _available = attr.ib(default=attr.Factory(list), validator=attr.validators.instance_of(list))
+    _avail_lock = attr.ib(default=threading.Lock())
+    _clients = attr.ib(default=attr.Factory(dict), validator=attr.validators.instance_of(dict))
+    _topics = attr.ib(default=attr.Factory(list), validator=attr.validators.instance_of(list))
+    _topic_lock = attr.ib(default=threading.Lock())
```
Comment (jluebbe) [inline labgrid/resource/mqtt.py:19]: You could probably use `factory=` for these. I'm not sure if maybe type annotations are a better solution for internal attributes.

File: labgrid/resource/mqtt.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+    def _create_mqtt_connection(self, host):
+        client = mqtt.Client()
+        client.connect(host)
+        client.on_connect = self._on_connect
+        client.on_message = self._on_message
+        client.loop_start()
+        return client
+
+    def on_resource_added(self, resource):
```
Comment (jluebbe) [inline labgrid/resource/mqtt.py:37]: Should this be private?

File: labgrid/driver/mqtt.py [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+        with self._status_lock:
+            self._status = status
+
+    def _on_connect(self, client, userdata, flags, rc):
+        client.subscribe(self.power.status_topic)
+
+    @Driver.check_active
+    @step()
+    def on(self):
+        self._client.publish(self.power.power_topic, payload="ON")
```
Comment (jluebbe) [inline labgrid/driver/mqtt.py:56]: Do we get feedback if this worked? When does this return?

File: labgrid/driver/mqtt.py [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+    def _get_status(self):
+        with self._status_lock:
+            status = self._status
+        return status
+
+    @Driver.check_active
+    @step()
+    def get(self):
+        self._client.publish(self.power.power_topic)
+        while self._status is None:
```
Comment (jluebbe) [inline labgrid/driver/mqtt.py:79]: Use util.timeout here.

File: labgrid/remote/exporter.py [hunk 7]
```diff
@@ -535,6 +535,23 @@ def _get_params(self):
 
 exports["LXAIOBusPIO"] = LXAIOBusNodeExport
 
+@attr.s(eq=False)
+class TasmotaPowerPortExport(ResourceExport):
+    """ResourceExport for LXAIOBusNode devices accessed via the HTTP API"""
```
Comment (jluebbe) [inline labgrid/remote/exporter.py:540]: The docstring is wrong.

File: labgrid/resource/mqtt.py [hunk 8]
```diff
@@ -0,0 +1,81 @@
+import queue
+import logging
+from time import monotonic
+
+import attr
+import paho.mqtt.client as mqtt
```
Comment (jluebbe) [inline labgrid/resource/mqtt.py:6]: This should also be imported only when needed.

File: labgrid/resource/mqtt.py [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+    _topics = attr.ib(default=attr.Factory(list), validator=attr.validators.instance_of(list))
+    _topic_lock = attr.ib(default=threading.Lock())
+    _last = attr.ib(default=0.0, validator=attr.validators.instance_of(float))
+
+    def __attrs_post_init__(self):
+        super().__attrs_post_init__()
+        self.log = logging.getLogger('MQTTManager')
+
+    def _on_connect(self, client, userdata, flags, rc):
+        client.subscribe("#")
```
Comment (jluebbe) [inline labgrid/resource/mqtt.py:27]: Is this a wildcard subscription? Shouldn't we subscribe to a more specific path?

File: labgrid/resource/mqtt.py [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+        host = resource.host
+        if host not in self._clients:
+            self._clients[host] = self._create_mqtt_connection(host)
+
+    def _on_message(self, client, userdata, msg):
+        payload = msg.payload.decode('utf-8')
+        topic = msg.topic
+        if payload == "Online":
+            with self._avail_lock:
+                self._available.append(topic)
```
Comment (jluebbe) [inline labgrid/resource/mqtt.py:47]: `_available` is a list but used like a set?

File: labgrid/resource/mqtt.py [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+        super().__attrs_post_init__()
+
+
+@target_factory.reg_resource
+@attr.s(eq=False)
+class TasmotaPowerPort(MQTTResource):
+    power_topic = attr.ib(default=None,
+                         validator=attr.validators.instance_of(str))
+    status_topic = attr.ib(default=None,
+                         validator=attr.validators.instance_of(str))
```
Comment (jluebbe) [inline labgrid/resource/mqtt.py:81]: Could we derive all these topic from a common prefix?

**Review by jluebbe — CHANGES_REQUESTED** [review]
We should add at least add a simple test which imports this, but better connect to a MQTT broker.

Maybe we don't need the split between NetworkTasmotaPowerPort and TasmotaPowerPort. This looks similar to the NetworkPowerPort and NetworkSerialPort, which are exported unmodified for external resources.

File: labgrid/driver/mqtt.py [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+
+
+@target_factory.reg_driver
+@attr.s(eq=False)
+class TasmotaPowerDriver(Driver, PowerProtocol):
+    bindings = {
+            "power": {"TasmotaPowerPort", "NetworkTasmotaPowerPort"}
+    }
+    delay = attr.ib(default=2.0, validator=attr.validators.instance_of(float))
+    _client = attr.ib(default=mqtt.Client(), validator=attr.validators.instance_of(mqtt.Client))
```
[the remark this answers is not in the document]
Comment (Emantor) [reply labgrid/driver/mqtt.py:26]: Since this is an `attr.ib` my expectation was that every class will get its own client.

File: labgrid/driver/mqtt.py (hunk 1 above)
Comment (Emantor) [reply labgrid/driver/mqtt.py:27]: `factory` is syntactic sugar for this same construct, but  I don't mind switching.

File: labgrid/resource/mqtt.py (hunk 3 above)
Comment (Emantor) [reply labgrid/resource/mqtt.py:19]: Same as above, but I can switch this to use `factory`.

File: labgrid/resource/mqtt.py (hunk 4 above)
Comment (Emantor) [reply labgrid/resource/mqtt.py:37]: This is called by the `ResourceManager` class, changing this will require changing all other classes using `ResourceManager` as well.

File: labgrid/driver/mqtt.py (hunk 5 above)
Comment (Emantor) [reply labgrid/driver/mqtt.py:56]: Yes, we should handle the returned class and wait for the topic to be delivered.

File: labgrid/resource/mqtt.py (hunk 11 above)
Comment (Emantor) [reply labgrid/resource/mqtt.py:81]: In theory yes, in practice we have already changed single topics inside our own infrastructure, which will make derivation from a simple token impossible. I implemented it this way at first, but everybody seems prone to change the topics on their own, so I'd rather  provide each topic separately.

File: labgrid/resource/mqtt.py (hunk 9 above)
Comment (Emantor) [reply labgrid/resource/mqtt.py:27]: We could subscribe only to the topics for whom we have registered resources.

File: labgrid/driver/mqtt.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (jluebbe) [reply labgrid/driver/mqtt.py:26]: No, with default= you set a single default value:
```
>>> import attr
>>> @attr.s()
... class Foo:
...   bar = attr.ib(default=dict())
... 
>>> a = Foo()
>>> b = Foo()
>>> a.bar is b.bar
True
>>> a.bar['this'] = 'that'
>>> b.bar
{'this': 'that'}
```

File: labgrid/driver/mqtt.py (hunk 1 above)
Comment (jluebbe) [reply labgrid/driver/mqtt.py:27]: Factory expects a function that should be called to produce the per-instance value.

File: labgrid/resource/mqtt.py (hunk 4 above)
Comment (jluebbe) [reply labgrid/resource/mqtt.py:37]: Ah, yes.

File: labgrid/resource/mqtt.py (hunk 11 above)
Comment (jluebbe) [reply labgrid/resource/mqtt.py:81]: Could you give an example of these topics?

File: labgrid/driver/mqtt.py [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+from ..protocol import PowerProtocol
+from ..step import step
+from ..util import Timeout
+
+
+@target_factory.reg_driver
+@attr.s(eq=False)
+class TasmotaPowerDriver(Driver, PowerProtocol):
+    bindings = {
+            "power": {"TasmotaPowerPort", "NetworkTasmotaPowerPort"}
```
Comment (Emantor) [inline labgrid/driver/mqtt.py:23]: We talked about this in a short call and agreed that the `NetworkTasmotaPortPort` does not make much sense. The exporter should export a plan `TasmotaPowerPort` instead, which gets instanciated again on the client side.

File: labgrid/resource/mqtt.py [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+import logging
+from time import monotonic
+
+import attr
+import paho.mqtt.client as mqtt
+import threading
+
+from .common import ManagedResource, ResourceManager, Resource
+from ..factory import target_factory
+from ..util import Timeout
```
Comment (Bastian-Krause) [inline labgrid/resource/mqtt.py:11]: Unused import.

File: labgrid/resource/mqtt.py [hunk 15]
```diff
@@ -0,0 +1,81 @@
+import queue
+import logging
+from time import monotonic
+
+import attr
+import paho.mqtt.client as mqtt
+import threading
```
Comment (Bastian-Krause) [inline labgrid/resource/mqtt.py:7]: Wrong import order, standard library first, then third-party, then local imports.

File: labgrid/resource/mqtt.py [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+
+from .common import ManagedResource, ResourceManager, Resource
```
Comment (Bastian-Krause) [inline labgrid/resource/mqtt.py:9]: Unused import: `Resource` is not used anywhere.

**Review by Bastian-Krause — DISMISSED** [review]
Some nitpicks :)

File: labgrid/resource/mqtt.py (hunk 11 above)
Comment (Emantor) [reply labgrid/resource/mqtt.py:79]: We could combine `status_topic` and `power_topic`, I'd like to keep the `avail_topic` apart since this is part of the MQTTResource and should work with other MQTT devices as well.

File: labgrid/driver/mqtt.py [hunk 17]
```diff
@@ -0,0 +1,95 @@
...
+    @Driver.check_active
+    @step()
+    def on(self):
+        self._publish(self.power.power_topic, "ON")
+
+
+    @Driver.check_active
+    @step()
+    def off(self):
+        self._publish(self.power.power_topic, "OFF")
```
Comment (Emantor) [inline labgrid/driver/mqtt.py:71]: We should wait for the message to be published and check that the status gets set via the get_status function.

File: labgrid/driver/mqtt.py [hunk 18]
```diff
@@ -0,0 +1,95 @@
...
+
+    def on_deactivate(self):
+        self._client.loop_stop()
+
+    def _on_message(self, client, userdata, msg):
+        if msg.payload == b'ON':
+            status = True
+        elif msg.payload == b'OFF':
+            status = False
+        with self._status_lock:
```
Comment (Emantor) [inline labgrid/driver/mqtt.py:48]: We don't need a `threading.Lock` for single python assignements or checks, so we can drop the lock from the driver here.

**Comment by Emantor** [comment]
I have adressed all review comments, the PR should be good to go.

File: labgrid/driver/mqtt.py [hunk 19]
```diff
@@ -0,0 +1,97 @@
...
+
+    @Driver.check_active
+    @step()
+    def get(self):
+        self._client.publish(self.power.power_topic)
+        timeout = Timeout(3.0)
+        while self._status is None:
+            time.sleep(0.1)
+            if timeout.expired:
+                raise MQTTError("Could not get initial status")
```
Comment (jluebbe) [inline labgrid/driver/mqtt.py:96]: Why is this the initial status?

File: labgrid/resource/lxaiobus.py [hunk 20]
```diff
@@ -10,6 +10,7 @@
 
 @attr.s(eq=False)
 class LXAIOBusNodeManager(ResourceManager):
+
```
Comment (jluebbe) [inline labgrid/resource/lxaiobus.py:13]: Spurious change.

File: labgrid/remote/exporter.py [hunk 21]
```diff
@@ -617,7 +617,6 @@ def _get_params(self):
 
 exports["LXAIOBusPIO"] = LXAIOBusNodeExport
 
-
```
Comment (jluebbe) [inline labgrid/remote/exporter.py:620]: Spurious change.

File: labgrid/resource/mqtt.py [hunk 22]
```diff
@@ -0,0 +1,76 @@
...
+    def _on_message(self, client, userdata, msg):
+        payload = msg.payload.decode('utf-8')
+        topic = msg.topic
+        if payload == "Online":
+            with self._avail_lock:
+                self._available.add(topic)
+        if payload == "Offline":
+            if topic in self._available:
+                with self._avail_lock:
+                    self._available.remove(topic)
```
Comment (jluebbe) [inline labgrid/resource/mqtt.py:46]: Use `.discard(topic)`.

**Review by jluebbe — CHANGES_REQUESTED** [review]

File: labgrid/driver/mqtt.py (hunk 19 above)
Comment (Emantor) [reply labgrid/driver/mqtt.py:96]: Since the initial status is unknown when running in the `labgrid-client` cli we need to request the topic and wait for the request to hit the message handler, therefore the initial state is `None`.

File: labgrid/resource/mqtt.py (hunk 22 above)
Comment (jluebbe) [reply labgrid/resource/mqtt.py]: Ah, sorry, i should have been more explicit: with `.remove()`, the initial check is no longer needed.

Comment (Emantor) [reply labgrid/resource/mqtt.py]: I guess you meant with `.discard()` the check for the object to be in the set is not required. I have adjusted the PR accordingly.

**Review by jluebbe — APPROVED** [review]
